Edits to retrievers section of Pipelines doc (#6530)
EnterpriseDB/docs: advocacy_docs/edb-postgres-ai/ai-accelerator/retrievers/concepts.mdx

---
title: "Retrievers concepts"
navTitle: "Concepts"
deepToC: true
description: "The concepts behind using retrievers in AI Accelerator Pipelines."
---

Retrievers can be used for semantic search or for retrieval augmented generation (RAG).
A configured retriever can compute embeddings (vectors) for your input data.
After the computing is done, the retriever offers interfaces to return data based on a query.

## Data sources and data types

A data source is where the data to be embedded/retrieved comes from. The aidb extension supports two types of data sources:

* Table &mdash; A column in a table in the PG database.
* Volume &mdash; A PGFS *volume*, which is a wrapper for accessing an S3 object store or local file system.

Data types are:

* Text
* Image

You can combine any data source with any data type:

* **Table and text**: The text, in this case, is just a column with TEXT/VARCHAR type in a Postgres table. An example is the description for a list of products.

> * **Table and image**: The image, in this case, would be stored in a BYTEA column in our table, which would hold the bytes of the image. Pipelines doesn't currently support large objects (also known as BLOB) types.

Reviewer: ```suggestion
* **Table and image**: The image, in this case, would be stored in a BYTEA column in the table, which would hold the bytes of the image. Pipelines doesn't currently support large objects (also known as BLOB) types.
```

---
title: "Retrievers concepts"
navTitle: "Concepts"
deepToC: true
description: "The concepts behind using retrievers in AI Accelerator Pipelines."
---

Retrievers can be used for semantic search or for retrieval augmented generation (RAG).
A configured retriever can compute embeddings (vectors) for your input data.

> After the computing is done, the retriever offers interfaces to return data based on a query.

Author: Why is there a level 2 heading called Concepts in a topic called Concepts? Should everything move up a level? If not, some text between the level2 and first level 3 head would be nice. And if so, I think the opening paragraph needs to maybe mention concepts? Like "These concepts will help you to understand how to approach configuring your retrievers." (Totally made up and I'm sure not accurate, but along those lines.

## Concepts

### Data sources and data types

A data source is where the data to be embedded/retrieved comes from. The aidb extension supports two types of data sources:

* Table &mdash; A column in a table in the PG database.

> * Volume &mdash; A PGFS *volume*, which is a wrapper for accessing an S3 object store or local filesystem.

Author: ```suggestion
* Volume &mdash; A PGFS *volume*, which is a wrapper for accessing an S3 object store or local file system.
```

Data types are:

* Text
* Image

You can combine any data source with any data type:

* **Table and text**: The text, in this case, is just a column with TEXT/VARCHAR type in a Postgres table. An example is the description for a list of products.

> * **Table and image**: The image, in this case, would be stored in a BYTEA column in our table, which would hold the bytes of the image. Pipelines doesn't currently support large objects (also known as BLOB) types.

Author: Querying use of first person here. "the" table?

* **Volume and text**: An example is text files in an S3 bucket.
* **Volume and image**: The images can be stored as objects in an S3 bucket.

### Embedding

> Retrievers always need a *vector table* to store the vector embeddings that belong to the source data. We use a model with *encoder* capabilities to generate the vectors for the source data.

Author: First person here: Should this be Pipelines in place of "we"?

Reviewer: ```suggestion
Retrievers always need a *vector table* to store the vector embeddings that belong to the source data. Pipelines uses a model with *encoder* capabilities to generate the vectors for the source data.
```

Pipelines creates the vector table and also supports using an existing one. The vector column dimensionality must fit the model.

### Consistency with source data

To retrieve correct and consistent data, embeddings must be in sync with the source data. In the case of the table data source, Pipelines uses triggers in Postgres to be notified about changes on the source table.

If the source table already has data/rows when the retriever is created, then a manual bulk embedding call must be made.

To keep the embeddings in sync going forward, you can activate auto-embedding.

### Bulk embedding and auto-embedding

Bulk embedding generates the embeddings for all the existing data in a source table. Use it to initialize the retriever if the source table isn't empty.

!!! Note
Bulk embedding doesn't sync the data. We recommend running it on an empty vector table.
!!!

> Auto-embedding can be manually enabled. It creates triggers in Postgres to keep the embeddings up to data if source data is added, updated, or removed.

Author: We generally don't refer to Postgres in this context as PG but spell it out, assuming that's what's meant here.

!!! Note
Auto-embedding currently works only with the table data source, not the volume source.
!!!

### Synchronous vs. Asynchronous embedding computation

* Bulk embedding is a blocking call that can’t be paused.

> * Auto-embedding runs on top of other queries during insert/update calls, while waiting on the embedding computation. This means insert/update queries on the source table will be slowed down.

Author: I couldn't imagine why "on top of" was quoted. If it needs explanation, i.e., you're using an analogy rather than something actual terminology, it would be better to explain what's meant by that.

Reviewer: ```suggestion
* Auto-embedding runs along with other queries during insert/update calls, while waiting on the embedding computation. This means insert/update queries on the source table will be slowed down.
```
